### ISSUE TITLE
Issue when running plugin in a multi-project build

### ISSUE THREAD

**Issue by wrobcio789** [EtienneMiret/sass-gradle-plugin#23]
Hello,

I'm using the plugin in 2 separate subprojects of a root project. These are: _libraries/uicomponents_ and _features/userdetails_. It's a JavaFX project and I'm adding a functionality to define different styles per screen. Each screen has its own subproject.

Right now, the plugin configuration in _build.gradle_ is the same in both subprojects:

```
jar {
    dependsOn compileSass
    from compileSass.outputDir
}

sass {
    noAutoCopy()
}

compileSass {
    sourceDir = project.file("${projectDir}/src/main/resources")
    outputDir = project.file("${buildDir}/resources/main")
    loadPath project.file("${rootDir}/common/sass-lib")
    sourceMap = none
}
```
However, when building the whole project there goes an error:

```
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :jar
> Task :assemble
> Task :compileTestJava NO-SOURCE
> Task :processTestResources NO-SOURCE
> Task :testClasses UP-TO-DATE
> Task :test NO-SOURCE
> Task :check UP-TO-DATE
> Task :build
> Task :libraries:core:compileJava
> Task :libraries:core:processResources NO-SOURCE
> Task :libraries:core:classes
> Task :libraries:core:jar
> Task :libraries:uicomponents:compileJava
> Task :libraries:uicomponents:processResources NO-SOURCE
> Task :libraries:uicomponents:classes
> Task :libraries:uicomponents:downloadSass UP-TO-DATE
> Task :libraries:uicomponents:installSass UP-TO-DATE
> Task :features:home:processResources
> Task :features:settings:processResources
> Task :features:userdetails:processResources
> Task :features:userdetails:downloadSass UP-TO-DATE

> Task :libraries:uicomponents:compileSass FAILED
The process cannot access the file because it is being used by another process.

> Task :features:userdetails:installSass
Execution optimizations have been disabled for task ':features:userdetails:installSass' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'C:\projects\javafx-app-template\.gradle\sass\archive\dart-sass-1.54.0-windows-x64.zip'. Reason: Task ':features:userdetails:installSass' uses this output of task ':libraries:uicomponents:downloadSass' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.3.3/userguide/validation_problems.html#implicit_dependency for more details about this problem.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.3.3/userguide/command_line_interface.html#sec:command_line_warnings

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.
14 actionable tasks: 9 executed, 5 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':libraries:uicomponents:compileSass'.
> A failure occurred while executing io.miret.etienne.gradle.sass.CompileSassWorkAction
   > Process 'command 'C:\projects\javafx-app-template\.gradle\sass\1.54.0\dart-sass\sass.bat'' finished with non-zero exit value 1
```

Everything is ok when running subprojects build separately.

I found a workaround for the problem and this is setting separate directory for the sass exec files per project, like this:
```
sass {
    directory = file("${rootDir}/.gradle/sass/${name}")
    noAutoCopy()
}
```
but when having a lot of subprojects I can imagine this being space consuming.

I'm using Gradle 7.3.3 version.

I would really appreciate some help on the matter :)

**Comment by EtienneMiret** [comment]
Thank-you for reporting this issue. I was able to reproduce it 😄.

However, I don’t really know how to fix it. The problem is that the plugin is instantiated once per subproject. Hence, you end up with two `downloadSass` and two `installSass` tasks, while there should be only one of each. And since gradle runs task in parallels, the `compileSass` task of one project may be ran at the same time as the `installSass` task of the other.

Your workaround ensure the sass executable is not overwritten while it is run. Another possible workaround is to set the maximum number of gradle workers to 1 (`./gradlew --max-workers=1` or `org.gradle.workers.max=1` in `gradle.properties`). This saves disk space at the cost of build time.

**Comment by wrobcio789** [comment]
Thank you for your time and the answer :)

I wish I could help with the issue, however I'm not that familiar with Gradle yet. I'll stick to the workaround for now.

**Comment by wrobcio789** [comment]
Ok, I took a deeper look into the specified tasks and tried to run copy of the plugin locally. I see that the Download task has the overwrite property set to false and I manually set the `DuplicateStrategy` of the `installSass` to [`Exclude`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/DuplicatesStrategy.html#EXCLUDE). It looks to me like this should work - second instance of `downloadSass` just skips downloading and second instance of `installSass` task should not modify or lock already present executable, which should cause already running compilation to finish.

But it doesn't, and I don't really understand why.

**Comment by olelond** [comment]
I'm facing this exact problem as well in a project with multiple subprojects using this plugin.
Unfortunately I'm no good with gradle either. I was reading the description in the link that gradle puts together with the error message: https://docs.gradle.org/7.5.1/userguide/validation_problems.html#implicit_dependency but I can't figure out if this presents any solution to our specific problem? 
Looking into the plugin code, it seems like you might be doing this correctly already @EtienneMiret - is that correct?
If this is the case there might be no generally accepted gradle way to to this. But that just seems strange because on should think the this kind of setup with muliple subprojects wouldn't be that unusual?

**Comment by EtienneMiret** [comment]
The issue we have is that this plugin creates a `downloadSass` task and an `installSass` task *per subproject*, while we need only one of each for the entire build.

The implicit dependency message is displayed because `:subprojectA:installSass` depends on the output file of `:subprojectB:downloadSass`, but only declares a dependency to `:subprojectA:downloadSass`. This, however, is just a symptom, not the issue.

Once I find time to do it (hopefully this weekend), I'll publish a new 1.4.2 release that will work around this issue by using a per subproject install path. This, however, **is just a workaround**. I’ll keep this issue open until I find a solution that avoid downloading sass once per subproject.

**Comment by EtienneMiret** [comment]
This bug will be properly fixed in release 1.4.3.

Note that you SHOULD now apply this plugin in the root project or, more precisely, in the latest common ancestor of all projects that make use of it.

That is, with the following structure:

```
:root
  :root:subproject-a
  :root:subproject-b
    :root:subproject-b:b1
    :root:subproject-b:b2
    :root:subproject-b:b3
  :root:c
```

Assuming `b1` and `b3` need this plugin, it SHOULD be applied to `subproject-b`. The `compileSass` task would still be configured in `b1` and `b3`, but the `sass` extension would now be configured in `subproject-b`.

Note that this is not a breaking change, as it is still possible to apply this plugin separately to `b1` and `b3`. But in that case, two `downloadSass` and two `installSass` task will be created, and they will download and install two separate copies of `dart-sass`.
